src/sentry/utils/sentryappwebhookrequests.py

```diff
@@ -0,0 +1,111 @@
+from __future__ import absolute_import
+
+import six
+import json
+
+from dateutil.parser import parse as parse_date
+from django.conf import settings
+from django.utils import timezone
+
+from sentry.utils import redis
+from sentry.models.sentryapp import VALID_EVENTS
+from sentry.models import Organization
+
+BUFFER_SIZE = 100
+
+
+class SentryAppWebhookRequestsBuffer(object):
+    """
+    Create a data structure to store basic information about Sentry App webhook requests in Redis
+    This should store the last 100 requests and last 100 errors (in different keys) for each event type, for each Sentry App
+    """
+
+    def __init__(self, sentry_app):
+        self.sentry_app = sentry_app
+
+        cluster_id = getattr(settings, "SENTRY_WEBHOOK_LOG_REDIS_CLUSTER", None)
+        if cluster_id is None:
+            self.client = redis.clusters.get("default").get_local_client(0)
+        else:
+            self.client = redis.redis_clusters.get(cluster_id)
+
+    def _convert_redis_request(self, redis_request, event):
+        """
+        Convert the request string stored in Redis to a python dict that can be returned from the errors endpoint
+        """
+        request = json.loads(redis_request)
+
+        formatted_request = {
+            "webhookUrl": request.get("webhook_url"),
+            "sentryAppSlug": self.sentry_app.slug,
+            "eventType": event,
+            "date": request.get("date"),
+            "responseCode": request.get("response_code"),
+        }
+
+        if "organization_id" in request:
+            org = Organization.objects.get(id=request["organization_id"])
+            formatted_request["organization"] = {"name": org.name, "slug": org.slug}
+
+        return formatted_request
+
+    def _add_to_buffer(self, buffer_key, item):
+        self.client.lpush(buffer_key, json.dumps(item))
+        self.client.ltrim(buffer_key, 0, BUFFER_SIZE - 1)
+
+    def add_request(self, response_code, org_id, event, url):
+        if event not in VALID_EVENTS:
+            return
+
+        request_key = self._get_redis_key(event)
+
+        time = timezone.now()
+        request_data = {
+            "date": six.binary_type(time),
+            "response_code": response_code,
+            "organization_id": org_id,
+            "webhook_url": url,
+        }
+
+        self._add_to_buffer(request_key, request_data)
+
+        # If it's an error add it to the error buffer
+        if not 200 <= response_code <= 299:
```

Comment:
@nolachen What about any 3xx status code?

Reply:
@scefali Should I be defining errors only as >400 status codes?